### PR TITLE
Move MDMerge to happen earlier in the build directly after Midl has ran.

### PR DESCRIPTION
**Pull request by jlaanstra**
This PR reverts back to previous behavior where we would generate the merged WinMD directly after the Midl target. This makes sure that project references work correctly earlier in the build process, which is important for razzle builds.

Notable changes include:
- Running MdMerge directly after the Midl target. This ensures that poject references are available early in the build process and is important for razzle builds.
- Move the generation of XamlMetadataProvider files to happen as part of the correct build steps. This means we include the idl file as part of the Midl targets and the cpp file as part of the ClCompile targets. This removes the need for some workarounds that exist in razzle today.
- Generate the component projection directly from the winmd files in the Merged folder instead of the output folder in case any other winmds are copied into the output folder.

### CODE REVIEW

File: src/package/cppwinrt/nuget/Microsoft.Windows.CppWinRT.targets [hunk 1]
```diff
@@ -243,19 +257,37 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
         <PropertyGroup>
-            <XamlMetaDataProviderIdl>
-                $(IntDir)XamlMetaDataProvider.idl
-            </XamlMetaDataProviderIdl>
-            <XamlMetaDataProviderCpp>
-                $(GeneratedFilesDir)XamlMetaDataProvider.cpp
-            </XamlMetaDataProviderCpp>
-            <_HasXaml>false</_HasXaml>
-            <_HasXaml Condition="'@(Page)' != '' Or '@(ApplicationDefinition)' != ''">true</_HasXaml>
+            <_DisableReferences>false</_DisableReferences>
```
Comment (jlaanstra) [inline src/package/cppwinrt/nuget/Microsoft.Windows.CppWinRT.targets]: @Scottj1s I am not 100% sure if I understood the _DisableReferences property. Can you double check that I refactored that the right way?

File: src/package/cppwinrt/nuget/Microsoft.Windows.CppWinRT.targets [hunk 2]
```diff
@@ -243,19 +257,39 @@ Copyright (C) Microsoft Corporation. All rights reserved.
         <Message Text="CppWinRTMdMergeMetadataDirectories: @(CppWinRTMdMergeMetadataDirectories)" Importance="$(CppWinRTVerbosity)"/>
     </Target>
 
-    <!--If building any Xaml content, include XamlMetaDataProvider in project metadata-->
-    <Target Name="CppWinRTAddXamlMetaDataProviderIdl" Condition="'$(XamlLanguage)' == 'CppWinRT' and '$(CppWinRTAddXamlMetaDataProviderIdl)' == 'true'">
+    <!-- Adds the XamlMetadataProvider idl to the Midl itemgroup, if building any xaml content -->
```
Comment (Scottj1s) [inline src/package/cppwinrt/nuget/Microsoft.Windows.CppWinRT.targets]: looks like the CppWinRTAddXamlMetaDataProviderIdl condition dropped out - that needs to apply to all 4 targets

File: src/package/cppwinrt/nuget/Microsoft.Windows.CppWinRT.targets (hunk 1 above)
Comment (Scottj1s) [reply src/package/cppwinrt/nuget/Microsoft.Windows.CppWinRT.targets]: I think so, but there's some clean up needed below (_DisableReferences references)

---
In reply to: [328949800](https://github.com/microsoft/xlang/pull/609#discussion_r328949800) [](ancestors = 328949800)

File: src/package/cppwinrt/nuget/Microsoft.Windows.CppWinRT.targets [hunk 3]
```diff
@@ -243,19 +257,39 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+                <CompilerIteration>XamlGenerated</CompilerIteration>
+            </ClCompile>
+        </ItemGroup>
+    </Target>
+
+    <!--If building any Xaml content, write XamlMetaDataProvider idl file -->
+    <Target Name="CppWinRTAddXamlMetaDataProviderIdl"
+            Condition="'@(Page)@(ApplicationDefinition)' != '' and '$(XamlLanguage)' == 'CppWinRT'">
+        <PropertyGroup>
             <_DisableReferences>false</_DisableReferences>
```
Comment (Scottj1s) [inline src/package/cppwinrt/nuget/Microsoft.Windows.CppWinRT.targets]: cruft?

File: src/package/cppwinrt/nuget/Microsoft.Windows.CppWinRT.targets [hunk 4]
```diff
@@ -33,15 +35,26 @@ Copyright (C) Microsoft Corporation. All rights reserved.
         <PrepareForBuildDependsOn>
             $(PrepareForBuildDependsOn);CppWinRTVerifyKitVersion
         </PrepareForBuildDependsOn>
+        <!-- Note: Before* targets run before Compute* targets. -->
```
Comment (Scottj1s) [inline src/package/cppwinrt/nuget/Microsoft.Windows.CppWinRT.targets]: with these changes in triggers, have you checked that all incremental rebuild scenarios still work?  (i.e., delete the xamlmetadataprovider.idl/cpp, touch single idl/cpp files in the project, etc)

Comment (jlaanstra) [reply src/package/cppwinrt/nuget/Microsoft.Windows.CppWinRT.targets]: I checked that incremental build didn't rebuild unnecessary files. I will also verify the scenario's you listed.

File: src/package/cppwinrt/nuget/Microsoft.Windows.CppWinRT.targets [hunk 5]
```diff
@@ -243,19 +257,39 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+                <CompilerIteration>XamlGenerated</CompilerIteration>
+            </ClCompile>
+        </ItemGroup>
+    </Target>
+
+    <!--If building any Xaml content, write XamlMetaDataProvider idl file -->
+    <Target Name="CppWinRTAddXamlMetaDataProviderIdl"
+            Condition="'@(Page)@(ApplicationDefinition)' != '' and '$(XamlLanguage)' == 'CppWinRT' and '$(CppWinRTAddXamlMetaDataProviderIdl)' == 'true'">
+        <PropertyGroup>
             <_DisableReferences>false</_DisableReferences>
```
Comment (Scottj1s) [inline src/package/cppwinrt/nuget/Microsoft.Windows.CppWinRT.targets]: still need to clean up _DisableReferences here

File: src/package/cppwinrt/nuget/Microsoft.Windows.CppWinRT.targets [hunk 6]
```diff
@@ -243,19 +257,39 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+            <Midl Include="$(XamlMetaDataProviderIdl)">
+                <DisableReferences Condition="$(_DisableReferences)">>true</DisableReferences>
+            </Midl>
+        </ItemGroup>
+    </Target>
+
+    <!-- Adds the XamlMetadataProvider cpp to the ClCompile itemgroup, if building any xaml content -->
+    <Target Name="CppWinRTComputeXamlGeneratedCompileInputs"
+            DependsOnTargets="$(CppWinRTComputeXamlGeneratedCompileInputsDependsOn)"
+            Condition="'@(Page)@(ApplicationDefinition)' != '' and '$(XamlLanguage)' == 'CppWinRT' and '$(CppWinRTAddXamlMetaDataProviderIdl)' == 'true'">
```
Comment (Scottj1s) [inline src/package/cppwinrt/nuget/Microsoft.Windows.CppWinRT.targets]: neat trick for combining && conditions

File: src/package/cppwinrt/nuget/Microsoft.Windows.CppWinRT.targets (hunk 4 above)
Comment (jlaanstra) [reply src/package/cppwinrt/nuget/Microsoft.Windows.CppWinRT.targets]: Delete xamlmetadataprovider.idl as well as touching files work correctly for the test solution.

File: src/package/cppwinrt/nuget/Microsoft.Windows.CppWinRT.targets (hunk 6 above)
Comment (jlaanstra) [reply src/package/cppwinrt/nuget/Microsoft.Windows.CppWinRT.targets]: Not my invention. This is how the xaml targets in the SDK do it :).

File: src/package/cppwinrt/nuget/Microsoft.Windows.CppWinRT.targets (hunk 5 above)
Comment (jlaanstra) [reply src/package/cppwinrt/nuget/Microsoft.Windows.CppWinRT.targets]: What exactly needs to be cleaned up here?

Comment (Scottj1s) [reply src/package/cppwinrt/nuget/Microsoft.Windows.CppWinRT.targets]: ah, never mind - I didn't see the use of it in XamlMarkupIdlImport

---
In reply to: [330668339](https://github.com/microsoft/xlang/pull/609#discussion_r330668339) [](ancestors = 330668339)

**Pull request merged**